### DISCUSSION TITLE
How to handle exceptions ?

### DISCUSSION THREAD

**Discussion by sumit-158** [fastapi/asyncer#375] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Asyncer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Asyncer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Asyncer but to [AnyIO](https://github.com/agronholm/anyio).
- [X] I already checked if it is not related to Asyncer but to [Trio](https://github.com/python-trio/trio).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncer
import logging
from typing import Union

from fastapi import FastAPI

async def func1(name: str):
    return {"message": f"Hello {name} form func1"}


async def func2():
    return {"message": f"Hello {name} form func2"}


@app.get("/test")
async def knowledge_panel(name_str: str):
    async with asyncer.create_task_group() as task_group:
        try:
            soon_value1 = task_group.soonify(func1)(name=name_str)
        except Exception:
            logging.exception()
        try:
            soon_value2 = task_group.soonify(func2)(name=name_str)
        except Exception:
            logging.exception()
    data = [soon_value1.value, soon_value2.value]
    return data
```

### Description

* How to handle an exception like here soon_value2 will throw an error, then how could I catch the exceptions and return user only single data rather than crashing the whole program?

### Operating System

Windows

### asyncer Version

0.0.1

### Python Version

Python 3.10.5

**Comment by tiangolo** [comment]
I assume you solved it, thanks for closing the issue 👍

**Comment by sumit-158** [comment]
@tiangolo not actually, I tried something else but it's not that robust and clean.

**Comment by gitierrez** [comment]
@sumit-158 how did you handle this? I'm having the exact same problem and can't find anything about it.

**Comment by sumit-158** [comment]
@gitierrez I used an async "exception_wrapper" decorator.
Here - https://github.com/openfoodfacts/facets-knowledge-panels/blob/6eda756eca7703f9cf0e1a7fad9cf868366df12d/app/exception_wrapper.py#L5

**Comment by MarkParker5** [comment, marked as answer]
I see there 3 options:
1. Handle exceptions inside func1 (func2)
```py

async def func1(name: str):
    try:
        return {"message": f"Hello {name} form func1"}
    except:
        return {} # fallback value

...
```

2. Add a decorator to wrap functions
```py
def catch(func):
    async def wrapped(*args, **kwargs):
        try:
            return await func(*args, **kwargs)
        except:
            return None
    return wrapped

# apply decorator
@catch
def func1(...):
    ...
    
def func2(...):
    ...
    
@app.get("/test")
async def knowledge_panel(name_str: str):
    async with asyncer.create_task_group() as task_group:
        soon_value1 = task_group.soonify(func1)(name=name_str) # catch applied to all func1
        soon_value2 = task_group.soonify(catch(func2))(name=name_str) # or wrap inly exact call
    data = [soon_value1.value, soon_value2.value] # don't forget to handle None case
    return data
```

3. Wrap entire group call
```py
@app.get("/test")
async def knowledge_panel(name_str: str):
    try:
        async with asyncer.create_task_group() as task_group:
            soon_value1 = task_group.soonify(func1)(name=name_str)
            soon_value2 = task_group.soonify(func2)(name=name_str)
        data = [soon_value1.value, soon_value2.value]
    except:
        data = []
    return data
```
